apollo-spaceport/src/lib.rs

```diff
@@ -0,0 +1,561 @@
+#[derive(Debug)]
+pub struct Reporter {
+    client: ReporterClient<Channel>,
+}
+
+impl Reporter {
+    /// Try to create a new reporter which will communicate with the supplied address.
+    ///
+    /// This can fail if:
+    ///  - the address cannot be parsed
+    ///  - the reporter can't connect to the address
+    pub async fn try_new<T: AsRef<str>>(addr: T) -> Result<Self, ReporterError>
+    where
+        prost::bytes::Bytes: From<T>,
+    {
+        let ep = Endpoint::from_shared(addr)?;
+        let client = ReporterClient::connect(ep).await?;
+        Ok(Self { client })
+    }
+
+    /// Try to create a new reporter which will communicate with the supplied address.
+    ///
+    /// This can fail if:
+    ///  - the address cannot be parsed
+    ///  - the reporter can't connect to the address
+    pub async fn try_new_with_static(addr: &'static str) -> Result<Self, ReporterError> {
+        let ep = Endpoint::from_static(addr);
+        let client = ReporterClient::connect(ep).await?;
+        Ok(Self { client })
+    }
+
+    /// Submit these stats onto the spaceport for eventual processing.
+    ///
+    /// The spaceport will buffer traces and stats, transferring them when convenient.
+    pub async fn submit_stats(
+        &mut self,
+        graph: ReporterGraph,
+        key: String,
+        stats: ContextualizedStats,
+    ) -> Result<Response<ReporterResponse>, Status> {
+        self.client
+            .add_stats(Request::new(ReporterStats {
+                graph: Some(graph),
+                key,
+                stats: Some(stats),
+            }))
+            .await
+    }
+
+    /// Submit this trace onto the spaceport for eventual processing.
+    ///
+    /// The spaceport will buffer traces and stats, transferring them when convenient.
+    pub async fn submit_trace(
+        &mut self,
+        graph: ReporterGraph,
+        key: String,
+        trace: Trace,
+    ) -> Result<Response<ReporterResponse>, Status> {
+        self.client
+            .add_trace(Request::new(ReporterTrace {
+                graph: Some(graph),
+                key,
+                trace: Some(trace),
+            }))
+            .await
+    }
+}
+
+/// The spaceport module contains the spaceport components
+pub mod spaceport {
+    use super::report;
+    use crate::{ReporterStats, ReporterTrace, TracesAndStats};
+    use bytes::BytesMut;
+    use flate2::write::GzEncoder;
+    use flate2::Compression;
+    use prost::Message;
+    use prost_types::Timestamp;
+    use reqwest::Client;
+    use std::collections::HashMap;
+    use std::io::Write;
+    use std::net::SocketAddr;
+    use std::sync::atomic::{AtomicU32, Ordering};
+    use std::sync::Arc;
+    use std::time::{SystemTime, UNIX_EPOCH};
+    use tokio::sync::mpsc::Sender;
+    use tokio::sync::Mutex;
+    use tokio::time::{interval, Duration, MissedTickBehavior};
+    use tonic::transport::{Error, Server};
+    use tonic::{Request, Response, Status};
+
```

Comment:
```suggestion
```

Reply:
Maybe I should have commented that more :thinking: 

As I discussed recently on our private slack channel, it's best to not try to make groups of imports at the moment. The reasoning is that these groups are subjective: I for one prefer to put the crate imports at the beginning for instance. Over time this becomes unmaintained because everybody who touch the code has no idea what is the actual grouping logic and it becomes a mix bag.

So for now please group all your `use` into a single paragraph (no newline) and let `cargo fmt` do its magic.

Reply:
Let's fix this in a follow-up PR and make `cargo fmt` just do the right three-pronged grouping that it supports now?

Reply:
> Let's fix this in a follow-up PR and make `cargo fmt` just do the right three-pronged grouping that it supports now?

I'm okay with that but in the meantime we should group all the imports together so it's easy with rustfmt and we don't end up with a mess and different rules everywhere.